### ISSUE TITLE
Keyboard layout changes not handled

### ISSUE THREAD

**Issue by kimci86** [SFML/SFML#895]
The issue #879 was closed but the issue of keyboard layout changes remains.

Using Ubuntu 14.04 and SFML 2.3, changing the keyboard layout while an application is running is not handled properly. Both KeyEvent and TextEntered events still follow the layout in effect when the application was launched.
Moreover, a warning message appears. (See [this comment](https://github.com/SFML/SFML/issues/879#issuecomment-101265811))

**Comment by binary1248** [comment]
Try out d25cb9c96c8ad8c6288b7eeaf9f78ae55422488f.

**Comment by kimci86** [comment]
I tried it. The warning message disappear but both KeyEvent and TextEntered events still follow the first layout.
By digging a little in the source, I found out that the `buildMap` functions from WindowImplX11.cpp and InputImpl.cpp are based on the `keysymMap` from Display.cpp. By making Display's `buildMap` function available to WindowImplX11 (by moving it to the sf::priv namespace) and calling it just before WindowImplX11's `buildMap` so that `keysymMap` gets updated, the WindowImplX11's `sfKeyMap` is updated correctly as well as InputImpl's `keycodeMap`. Then, the KeyEvent events contain the right Keyboard::Key value and isKeyPressed behaves correctly.

To summarize, Display.cpp `buildMap` function must be called first.
However, TextEntered events still follow the first layout.

Besides, reading the source, I felt that the way InputImpl's `keycodeMap` is built is somewhat brutal. Moreover, I noticed that several map rebuild are triggered for a single keyboard layout change (up to 22 times if I plug my keyboard in). If I understand correctly, building `keycodeMap` could be done in a single loop by "inverting" WindowImplX11's `sfKeyMap` (even if it is not exactly bijective), but maybe this is a little off-topic.

**Comment by binary1248** [comment]
Try out c94536ac631da364549f541066b2484287707a72.

**Comment by kimci86** [comment]
This is not working, but I noticed that `xkbType` is always 0.
I replaced `if (xkbType == XCB_XKB_MAP_NOTIFY)`
by `if (xkbType == XCB_XKB_NEW_KEYBOARD_NOTIFY || xkbType == XCB_XKB_MAP_NOTIFY)`
and everything gets updated as intended. I don't know if this would be valid for every system though.
As before, TextEntered events follow the first layout.

**Comment by binary1248** [comment]
Define "not working". What are you doing? Changing the keyboard layout or re-plugging the device?

**Comment by abwrts** [comment]
Hey, guys. I make a program on c++ like billiard. So I want to ask for some advice about physics for balls and its collision with each other.

**Comment by kimci86** [comment]
> Define "not working".

I mean that the commit c94536a does not handle keyboard layout changes. Nothing is updated when I switch keyboard layout because this if condition never pass: `if (xkbType == XCB_XKB_MAP_NOTIFY)`

> What are you doing? Changing the keyboard layout or re-plugging the device?

Yesterday I was only changing the keyboard layout. As surprising as it may seem, `xkbType` is always equal to 0 (that is to say `XCB_XKB_NEW_KEYBOARD_NOTIFY`). Today I also tested re-plugging the keyboard and `xkbType` is also 0.

**Comment by binary1248** [comment]
a5de801c37ea9498dc5ed3a99201be8c2668217f should do the trick. Was a bit more involved than I originally thought.

**Comment by kimci86** [comment]
Now, KeyEvent and TextEvent data follow the new layout. :clap:
For me, this commit solves the issue.

**Comment by eXpl0it3r** [comment]
Fixed in 0df1c97af741bd1659780ce8c150ed8d357e9a4e on branch 2.3.x
